src/Altinn.App.Core/Features/Correspondence/Models/CorrespondenceRequest.cs

```diff
@@ -0,0 +1,238 @@
+using System.ComponentModel.DataAnnotations;
+using System.Diagnostics.CodeAnalysis;
+using Altinn.App.Core.Features.Correspondence.Exceptions;
+using Altinn.App.Core.Models;
+
+namespace Altinn.App.Core.Features.Correspondence.Models;
+
+/// <summary>
+/// Base functionality for correspondence models
+/// </summary>
+public abstract record CorrespondenceBase
+{
+    internal static void AddRequired(MultipartFormDataContent content, string value, string name)
+    {
+        if (string.IsNullOrWhiteSpace(value))
+            throw new CorrespondenceValueException($"Required value is missing: {name}");
+
+        content.Add(new StringContent(value), name);
+    }
+
+    internal static void AddRequired(MultipartFormDataContent content, Stream data, string name, string filename)
+    {
+        if (data is null)
+            throw new CorrespondenceValueException($"Required value is missing: {name}");
+
+        content.Add(new StreamContent(data), name, filename);
+    }
+
+    internal static void AddIfNotNull(MultipartFormDataContent content, string? value, string name)
+    {
+        if (!string.IsNullOrWhiteSpace(value))
+            content.Add(new StringContent(value), name);
+    }
+
+    internal static void AddListItems<T>(
+        MultipartFormDataContent content,
+        IReadOnlyList<T>? items,
+        Func<T, string> valueFactory,
+        Func<int, string> keyFactory
+    )
+    {
+        if (IsEmptyCollection(items))
+            return;
+
+        for (int i = 0; i < items.Count; i++)
+        {
+            string key = keyFactory.Invoke(i);
+            string value = valueFactory.Invoke(items[i]);
+            content.Add(new StringContent(value), key);
+        }
+    }
+
+    internal static void SerializeListItems<T>(MultipartFormDataContent content, IReadOnlyList<T>? items)
+        where T : ICorrespondenceItemSerializer
+    {
+        if (IsEmptyCollection(items))
+            return;
+
+        // Ensure unique filenames for attachments
+        if (items is IReadOnlyList<CorrespondenceAttachment> attachments)
+        {
+            var hasDuplicateFilenames = attachments
+                .GroupBy(x => x.Filename.ToLowerInvariant())
+                .Where(x => x.Count() > 1)
+                .Select(x => x.ToList());
+
+            foreach (var duplicates in hasDuplicateFilenames)
+            {
+                for (int i = 0; i < duplicates.Count; i++)
+                {
+                    duplicates[i].FilenameClashUniqueId = i + 1;
```

Comment:
Good point on the testing of this, afaik there is zero coverage on that. Like we discussed briefly the other day, this particular implementation still doesn't make me completely at ease.

I remember doing it this way to avoid pushing too much data from the top-level `.Serialize()` method down to the attachment object, but I think perhaps this would be the correct approach. The serializer deals out new filenames directly to nested serialize models as required, but that state never gets stored in the records themselves.